File: python/04_011.py
class UserService:
  """
  Service for user operations.
  """
  def __init__(self, user_repository):
     self.user_repository = user_repository
  
  def get_user(self, username):
     return self.user_repository.find_by_username(
        username)
  
  def create_user(self, user):
     if self.user_repository.exists_by_username(
           user.get_username()):
        raise UserAlreadyExistsException()
     self.user_repository.save(user)

class EmailService:
  """
  Service for email operations.
  """
  def send_welcome_email(self, email):
     # Send welcome email
     pass

File: cpp/04_011.cpp
#include <string>
#include <memory>
#include <stdexcept>

class User; // Forward declaration
class UserRepository; // Forward declaration
class UserAlreadyExistsException; // Forward declaration

class UserService {
private:
    std::shared_ptr<UserRepository> userRepository;
    
public:
    UserService(std::shared_ptr<UserRepository> repository) 
        : userRepository(repository) {}
 
    User getUser(const std::string& username) {
        return userRepository->findByUsername(username);
    }
 
    void createUser(const User& user) {
        if (userRepository->existsByUsername(
                user.getUsername())) {
            throw UserAlreadyExistsException();
        }
        userRepository->save(user);
    }
};

class EmailService {
public:
    void sendWelcomeEmail(const std::string& email) {
        // Send welcome email
    }
};